### ISSUE TITLE
[macOS] 'Move Data File' should execute on separate thread and give the user some feedback

### ISSUE THREAD

**Issue by beeb** [transmission/transmission#108]
When moving the data for a torrent via the contextual menu or 'File->Move Data File To...', after selecting the destination folder, the UI freezes during the whole moving process, because this is done on the main thread (usually reserved for simple/fast operations and UI updates).

What's more, no feedback regarding the progress of the operation is given to the user. Depending on the writing speed of the destination disk and the file size, this could last anywhere from a few seconds to dozens of minutes. The user should have a progress indicator displayed and the operation should be done on a separate thread in order to not freeze the whole UI (even if this means making the progress window 'modal' or a cocoa 'sheet' attached to the main window, to disallow further UI interaction).

A novice user might get scared if the program hangs in any way, especially without feedback.

EDIT: found this in the source inside`Torrent.m`... :/

    while (status == TR_LOC_MOVING) //block while moving (for now)
        [NSThread sleepForTimeInterval: 0.05];

**Comment by mikedld** [comment]
Related to [TRAC-1753](https://trac.transmissionbt.com/ticket/1753).

**Comment by beeb** [comment]
It's from 8 years ago hahaha 😂 longest-running bug in history?

**Comment by Dulanic** [comment]
Would love to see this fixed, this can cause a significant "freeze" when moving large amount of data. This is unrelated to MacOS also. It happens /w linux too.

**Comment by rodibin-at-placemaildotonline** [comment]
This freezes on android transmission's too would love to see this fixed soon hopefully don't have to wait 7 more years :(

**Comment by thekiefs** [comment]
Happening on Android for me too

200GB file freezes the UI for 10min+

**Comment by Dulanic** [comment]
Honestly guys, if it's a concern for you switch. Other clients don't have this problem. Wish I was better at coding to help, but qbittorrent doesn't have this problem.

**Comment by bagobones** [comment]
The UI / API being blocked by file operations continues to get worse as torrents get larger. This is really the only major issue I have with transmission.. I love just about everything else bout it.

**Comment by chrisjohnson00** [comment]
Bump!
I'm running Transmission on Kubernetes, and when this issue surfaces, the heath/liveness checks fail and the container is restarted.  When this happens, I need to manually move files from "incomplete" to "download".

**Comment by dantebarba** [comment]
Same here. Moving to a cloud drive it's impossible, it completely freezes the interface.

**Comment by Yonn-Trimoreau** [comment]
You should move to another torrent client if this is important to you. This will never be fixed. Deluge does not have this issue for example.

**Comment by beeb** [comment]
> This will never be fixed.

In which authority do you speak?

**Comment by Yonn-Trimoreau** [comment]
I'm just taking things realistically:
- 12 years after the original ticket has been created, still not resolved and not even an indication that is will ever be fixed.
- Only 11 comments and 14 likes on this GitHub issue, after 4 years of gathering interest.

 But pardon me, I have no authority on this matter and I should have said:

> This **_almost certainly_** will never be fixed

**Comment by beeb** [comment]
One can hope! The issue picked up some speed recently, sometimes that's all you need.

I was genuinely curious if you were involved in the project and wanted to mark this as "won't fix".

**Comment by bagobones** [comment]
@Yonn-Trimoreau if you prefer other clients then go comment on their issue trackers.. While we all know there are issues with transmission your comments really do not add anything to this issue thread.. We can all see how long it has been open. Development really took a long pause and now that there is new interest, there is a good opportunity for fresh eyes to look at these issues.

**Comment by dantebarba** [comment]
> You should move to another torrent client if this is important to you. This will never be fixed. Deluge does not have this issue for example.

I currently use transmission and unfortunately I invested a lot of time developing tools to fulfill my use case. I'll probably have to switch though.

**Comment by andreygursky** [comment]
@Yonn-Trimoreau:
> Only 11 comments and 14 likes on this GitHub issue, after 4 years of gathering interest.

There is another issue #694 with some more comments, details and likes.

**Comment by Yonn-Trimoreau** [comment]
@bagobones I do not prefer other clients. In fact, I went away from Transmission with regrets because of this only issue. All the other features Transmission proposes are super duper cool and fits my needs.
I did not know about the "new interest" happening. Nice to hear that. Hope this ticket will be fixed then.

**Comment by amcgregor** [comment]
I've been dealing with multi-hundred-gigabyte archives, totalling several dozen terabytes.  (Talking a complete copy of all StackExchange sites, complete copy of Wikipedia and Wikimedia sites, Project Gutenberg, "leaked passwords" PCHT, CrashCourse video content, archives of government research FTP sites, …)  I've been forced to add all torrents paused as moving them after initial file allocation is simply infeasible.  Transfers appear to lock up Tarnsmission.app with partially rendered UI until the operation completes:

<img width="718" alt="Screen Shot 2021-01-30 at 09 21 06" src="https://user-images.githubusercontent.com/76750/106359082-4122e600-62de-11eb-83a7-d06f45bf1c36.png">

Even getting that screenshot involved only mouse operations via the Screenshot application. Global screen capture shortcuts fail as long as the operation is underway.

### Problems

* File movement happens imperatively, **within the main application thread** at the point of user confirmation.

* File movement is **not** a **background operation** or state similar to "verifying integrity", _when it should be._  E.g. a "relocating" state.

* This ongoing operation, not allowing the main thread to respond to SystemUIServer or other application process events, prevents freeing of FirstResponder back to the system.

That last item has **catastrophic** consequences for keyboard input across the entire system—all applications are prevented from "gaining focus" and accepting keyboard entry.  Example situation, Safari address bar, every keypress results in a system error beep. Because every keypress is _actually being routed to Transmission_.

This might look focused, but it is **not**:

<img width="738" alt="Screen Shot 2021-01-30 at 09 24 41" src="https://user-images.githubusercontent.com/76750/106359123-93fc9d80-62de-11eb-8550-8ca35578d20c.png">

So, yup, temporary workaround: disable automatic activation of newly added torrents.  Move them where they need to be, then activate them manually.

**Comment by thekiefs** [comment]
Related to https://github.com/transmission/transmission/issues/1314 , https://github.com/transmission/transmission/issues/944

Seems like it might be fixed by https://github.com/transmission/transmission/pull/1092

**Comment by RobCrowston** [comment]
#1092 will not fix this hang. In some cases it may make transfers faster.

I have a half written, separate patch that _will_ fix this freezing behaviour, but to be frank, if a simple change like #1092 cannot be merged, I do not imagine a more complicated patch to separate file moves into a dedicated thread would make the cut either.

**Comment by quexten** [comment]
@RobCrowston Thank you for your work on #1092. Now that it has been merged, could you take another look if you can submit a merge request for this one too (since you mentioned you had worked on it)? It would be a big improvement.

**Comment by ckerr** [comment]
Closing as a dupe of https://github.com/transmission/transmission/issues/2462. I realize this ticket was filed before 2462, but there are so many overlapping tickets at this point that it makes sense to start with a new "big picture" ticket.

**Comment by ckirk** [comment]
**Decent workaround:** just move the files manually (outside of Transmission) then relink back in Transmission.

To Relink:
- Right click on file > Move Data File To...
- Select new file location
- Resume the torrent and it should go green

Even works with multiple torrents selected and accomplishes the same thing without freezing your entire computer for the duration of the file transfer. Haven't tested this with incomplete torrents, but works fine with completed ones.

**Comment by beankylla** [comment]
also facing the issue and got curious :) and issue is already here. don't know if i can help without coding skills but am here :)

**Comment by beankylla** [comment]
also facing the issue and got curious :) and issue is already here. don't know if i can help without coding skills but am here :)

**Comment by a4vg** [comment]
I'm using Transmission on a Raspberry Pi with an slow SD card and this issue keeps bothering me event with small files (an 800MB file for example). Each time a torrent finishes the whole Web UI hangs for a few minutes.

**Comment by amcgregor** [comment]
> I'm using Transmission on a Raspberry Pi with an slow SD card and this issue keeps bothering me even with small files (an 800MB file for example). Each time a torrent finishes the whole Web UI hangs for a few minutes.

That's a particular problem if your "staging area" (in-progress downloads) are on one volume, and the target location to download to is on a **different** volume. You can avoid this by ensuring your in-progress download is saved to the destination from the start, or by having your in-progress location on the same volume as the destination. Moving-into-place on completion is a trivial (and _fast_) file rename operation in that scenario.

In cases where I'm not prompted when initially adding (i.e. magnet links) I take the hit by manually transferring once it has sufficient information to allocate files. Pause, manual move outside of Transmission, "move…" within Transmission to where I copied it to, verify integrity, re-start. It's a tiny bit tedious, but likely faster and with no application stalling this way.

At least some of my other media pipelines can still benefit from an SSD during ingest. 😜
